generators/app/templates/src/services/utils.ts

```diff
@@ -0,0 +1,131 @@
+  values: P;
+  request: Request<P, D, E>;
+  onPrefetch: () => void;
+  onSuccess: Success<D>;
+  onError: Failure<E>;
+  onPostFetch: PostFetch<D, E>;
+}
+
+const executeAsyncRequest = async <P, D, E>({
+  values,
+  request,
+  onPrefetch,
+  onSuccess,
+  onError,
+  onPostFetch
+}: AsyncRequest<P, D, E>) => {
+  onPrefetch();
+  const response = await request(values);
+  if (response.ok) {
+    onSuccess(response.data);
+  } else {
+    onError({ problem: response.problem, errorData: response.data });
+  }
+  onPostFetch(response);
+};
+
+// Returns a request to execute manually at some point, and the variables that will be updated when it does
+export const useLazyRequest = <P, D, E>({
+  request,
+  withPostSuccess,
+  withPostFailure,
+  initialState = null,
+  withPostFetch,
+  transformResponse = response => response
+}: AsyncRequestHookParams<P, D, E>): [Nullable<D>, boolean, Nullable<Error<E>>, (params: P) => void] => {
+  const [state, setState] = useState(initialState);
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<Nullable<Error<E>>>(null);
+
+  const sendRequest = useCallback(
+    values => {
+      executeAsyncRequest({
+        values,
+        request,
+        onPrefetch: () => {
+          setLoading(true);
+          setState(initialState);
+          setError(null);
+        },
+        onSuccess: data => {
+          const transformedResponse = data ? transformResponse(data) : undefined;
+          setState(transformedResponse);
+          withPostSuccess?.(transformedResponse);
+        },
+        onError: errorInfo => {
+          setError(() => errorInfo);
+          withPostFailure?.(errorInfo);
+        },
+        onPostFetch: response => {
+          setLoading(false);
+          if (response.data) {
+            withPostFetch?.(transformResponse(response.data));
+          }
+        }
+      });
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [initialState, request, withPostFailure, withPostSuccess]
+  );
+
+  return [state, loading, error, sendRequest];
+};
+
+// Executes a request each time a dependency changes and returns the values and the refetch function
+// in case you want to execute it again
+export const useRequest = <P, D, E>(
+  {
+    request,
+    payload,
+    withPostSuccess,
+    withPostFailure,
+    initialState = null,
+    withPostFetch,
+    transformResponse = response => response
+  }: AsyncRequestHookParamsWithPayload<P, D, E>,
+  dependencies: []
+): [Nullable<D>, boolean, Nullable<Error<E>>, (params: P) => void] => {
+  const [state, loading, error, sendRequest] = useLazyRequest({
+    request,
+    withPostSuccess,
+    withPostFailure,
+    initialState,
+    withPostFetch,
+    transformResponse
+  });
+
+  useEffect(
+    () => {
+      sendRequest(payload);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    dependencies
+  );
+
+  return [state, loading, error, sendRequest];
+};
```

Comment:
I think these 2 should be on a folder call `Hooks` where you can create your custom hooks. I would also add it to the main src and add an access with `~`
```
components
service
hooks
  | useRequest
     | index.ts
     | constants.ts (in case you need it)
     | test.ts
  | useAsyncRequest
....
```

Reply:
Im saying this because we should encourage devs to create their custom hooks which not only is a good practice, it solves lots of problems of code repetition. I think with this we normalize the place to where create them and we already give them 2 examples of how to create custom hooks

Reply:
I like it, I'll change it